server/src/com/thoughtworks/go/config/GoPartialConfig.java

```diff
@@ -12,108 +12,166 @@
+        } catch (Exception e) {
+            if (repoConfig != null) {
+                String description = String.format("%s -  Config-Repo: %s", GoConfigValidity.invalid(e).errorMessage(), newPart.getOrigin().displayName());
+                ServerHealthState state = ServerHealthState.error(INVALID_CRUISE_CONFIG_MERGE, description, HealthStateType.general(HealthStateScope.forPartialConfigRepo(repoConfig)));
+                serverHealthService.update(state);
+            }
+            return false;
+        }
+    }
+
+    private PartialConfig findMatchingPartial(CruiseConfig cruiseConfig, String fingerprint) {
+        PartialConfig matchingPartial = findMatchingPartial(cruiseConfig, fingerprint, cachedGoPartials.getValid(fingerprint));
+        if (matchingPartial == null) {
+            matchingPartial = findMatchingPartial(cruiseConfig, fingerprint, cachedGoPartials.getKnown(fingerprint));
+        }
+        return matchingPartial;
+    }
+
+    private PartialConfig findMatchingPartial(CruiseConfig cruiseConfig, String fingerprint, PartialConfig partial) {
+        PartialConfig matching = null;
+        if (partial != null) {
+            for (PartialConfig partialConfig : cruiseConfig.getPartials()) {
+                if (partialConfig.getOrigin() instanceof RepoConfigOrigin && (((RepoConfigOrigin) partialConfig.getOrigin()).getMaterial().getFingerprint().equals(fingerprint))) {
+                    matching = partialConfig;
+                    break;
+                }
+            }
+        }
+        return matching;
     }
 
     @Override
-    public synchronized void onChangedRepoConfigWatchList(ConfigReposConfig newConfigRepos) {
-        boolean removedAny = false;
+    public void onChangedRepoConfigWatchList(ConfigReposConfig newConfigRepos) {
+        boolean removed = false;
         // remove partial configs from map which are no longer on the list
-        for(String fingerprint : this.fingerprintToLatestValidConfigMap.keySet())
-        {
-            if(!newConfigRepos.hasMaterialWithFingerprint(fingerprint))
-            {
-                this.fingerprintToLatestValidConfigMap.remove(fingerprint);
-                removedAny = true;
+        for (String fingerprint : cachedGoPartials.getFingerprintToLatestKnownConfigMap().keySet()) {
+            if (!newConfigRepos.hasMaterialWithFingerprint(fingerprint)) {
+                removed = true;
+                cachedGoPartials.removeKnown(fingerprint);
             }
         }
-        //fire event about changed partials collection
-        if(removedAny)
-            this.notifyListeners();
+        for (String fingerprint : cachedGoPartials.getFingerprintToLatestValidConfigMap().keySet()) {
+            if (!newConfigRepos.hasMaterialWithFingerprint(fingerprint)) {
+                removed = true;
+                cachedGoPartials.removeValid(fingerprint);
+            }
+        }
+        if (removed) {
+            updateConfig(null, null, null);
```

Comment:
No, I haven't thought through this yet :)
May be I should have mentioned this in the PR description. There are few things which do not work as one would expect as far as config-repo goes
* Deletion of config-repos. As of now this will cause problems if end-users remove a partial which was being referred to from another partial. 
* Surviving a Go server restart: A server restart causes Go to forget all known and valid partials. After the next MDU, the server finds all the known partials. However, if the latest set of known partials are invalid, the valid partial list will be empty. We do not persist this info in db as of now.
* various config APIs and their call backs: We need to ensure the same two step validation happen for the config APIs too. Also, revisit the listeners to check if we need to add any additional logic to handle remote environments/groups.